From 2e0bf4a557cd997fb659f5f66fd2938617250e7f Mon Sep 17 00:00:00 2001
From: Subrata Chatterjee <schatterjee@habana.ai>
Date: Mon, 18 Jul 2022 13:56:43 +0300
Subject: [PATCH] [SW-95911] gaudi2: bist: remove engine sram bists

sram bist to be run after PLLs are configured to functional clock.
PLL configurations are to be done in preboot. So bist for engine sram
bist will be moved to preboot after pll configurations are done.
This patch removes those sram bist from ppboot. Only sram bist for 48MB
SRAM is kept. Preboot runs from this 48MB SRAM, so bist on this memory
should be run before. SRAM PLLs to be configured in a separate patch.

Signed-off-by: Subrata Chatterjee <schatterjee@habana.ai>
Change-Id: I942f2ffa0d5a37f364bf56f86397f66d1e8c86f0
---

diff --git a/arch/gaudi2/arc/asic.c b/arch/gaudi2/arc/asic.c
index d2a980f..2430a35 100644
--- a/arch/gaudi2/arc/asic.c
+++ b/arch/gaudi2/arc/asic.c
@@ -31,90 +31,53 @@
 
 #define SDRAM_BASE_FULL 0x1000FFFFFD000000ull
 
-#define	DELAY_100K_USEC			100000
 #define	DELAY_5K_USEC			5000
 #define	DELAY_67_USEC			67
-#define DELAY_7P5K_USEC			7500
 #define DELAY_1K_USEC			1000
-#define DELAY_17_USEC			17
-#define DELAY_8K_USEC			8000
-#define DELAY_57_USEC			57
-#define DELAY_4K_USEC			4000
-#define DELAY_20_USEC			20
-#define DELAY_104_USEC			104
-#define DELAY_36_USEC			36
-#define DELAY_44_USEC			44
-#define DELAY_55_USEC			55
-
-#define GAUDI2_DFX_NUM			5
-#define GAUDI2_MAX_RING_NUM		28
+#define GAUDI2_SRAM_RINGS_NUM		32
 
 struct hl_ras_sram_bist_data		*bist_data;
 
 const static uint8_t dfx_value_table[] = { 0x27, 0x2b, 0x2f, 0x33, 0x37 };
-const static uint8_t ring_count_table[] = { 28, 26, 15, 15, 22 };
 
-const static uint8_t gauid2_ring_type_table[GAUDI2_DFX_NUM][GAUDI2_MAX_RING_NUM] = {
-	{ /* dfx = 0 */
-		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
-		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_SOB_XDMA, SRAM_RING_TYPE_VDEC,
-		SRAM_RING_TYPE_VDEC, SRAM_RING_TYPE_VDEC, SRAM_RING_TYPE_VDEC,
-		SRAM_RING_TYPE_MME, SRAM_RING_TYPE_HMMU_XBAR, SRAM_RING_TYPE_EDMA1_HMMU3,
-		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
-		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
-		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_NIC,
-		SRAM_RING_TYPE_HMMU_2, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_XBAR_MID
-	},
-	{ /* dfx = 1 */
-		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
-		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_VDEC, SRAM_RING_TYPE_VDEC,
-		SRAM_RING_TYPE_MME, SRAM_RING_TYPE_HMMU_XBAR, SRAM_RING_TYPE_EDMA1_HMMU3,
-		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
-		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
-		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_XBAR_MID,
-		SRAM_RING_TYPE_HMMU_2, SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC,
-		SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_MAX
-	},
-	{ /* dfx = 2 */
-		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
-		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
-		SRAM_RING_TYPE_SOB_ROT, SRAM_RING_TYPE_VDEC, SRAM_RING_TYPE_VDEC,
-		SRAM_RING_TYPE_HMMU_XBAR, SRAM_RING_TYPE_HMMU_MC, SRAM_RING_TYPE_XBAR_MID,
-		SRAM_RING_TYPE_HMMU_2, SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_MAX
-	},
-	{ /* dfx = 3 */
-		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
-		SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC, SRAM_RING_TYPE_TPC,
-		SRAM_RING_TYPE_SOB_ROT, SRAM_RING_TYPE_VDEC, SRAM_RING_TYPE_VDEC,
-		SRAM_RING_TYPE_HMMU_XBAR, SRAM_RING_TYPE_HMMU_MC, SRAM_RING_TYPE_XBAR_MID,
-		SRAM_RING_TYPE_HMMU_2, SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_MAX
-	}, /* dfx = 4 */
-	{
-		SRAM_RING_TYPE_MME, SRAM_RING_TYPE_MME, SRAM_RING_TYPE_SRAM,
-		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
-		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
-		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
-		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
-		SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM, SRAM_RING_TYPE_SRAM,
-		SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC, SRAM_RING_TYPE_NIC,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX, SRAM_RING_TYPE_MAX,
-		SRAM_RING_TYPE_MAX
-	},
+struct gaudi2_sram_ring_info {
+		uint8_t dfx;
+		uint8_t ring;
+};
+
+const static struct gaudi2_sram_ring_info gaudi2_sram_rings[] = {
+							{0, 16}, /* D0R16 */
+							{0, 17}, /* D0R17 */
+							{0, 18}, /* D0R18 */
+							{0, 19}, /* D0R19 */
+							{0, 20}, /* D0R20 */
+							{0, 21}, /* D0R21 */
+							{0, 22}, /* D0R22 */
+							{0, 23}, /* D0R23 */
+							{1, 13}, /* D1R13 */
+							{1, 14}, /* D1R14 */
+							{1, 15}, /* D1R15 */
+							{1, 16}, /* D1R16 */
+							{1, 17}, /* D1R17 */
+							{1, 18}, /* D1R18 */
+							{1, 19}, /* D1R19 */
+							{1, 20}, /* D1R20 */
+							{4, 3},  /* D4R3 */
+							{4, 4},  /* D4R4 */
+							{4, 5},  /* D4R5 */
+							{4, 6},  /* D4R6 */
+							{4, 7},  /* D4R7 */
+							{4, 8},  /* D4R8 */
+							{4, 9},  /* D4R9 */
+							{4, 10}, /* D4R10 */
+							{4, 11}, /* D4R11 */
+							{4, 12}, /* D4R12 */
+							{4, 13}, /* D4R13 */
+							{4, 14}, /* D4R14 */
+							{4, 15}, /* D4R15 */
+							{4, 16}, /* D4R16 */
+							{4, 17}, /* D4R17 */
+							{4, 18}, /* D4R18 */
 };
 
 static struct hl_pll_cfg gaudi2_arch_pll_cfg[] = {
@@ -206,162 +169,6 @@
 	return gaudi2_sms_check_pattern(val, ones_mask, zeros_mask);
 }
 
-static int gaudi2_tpc_ring_pattern(enum gaudi2_sram_self_repair_op op)
-{
-	uint32_t expected_ones, expected_zeros;
-	int err, ret = 0;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL23, mmPSOC_DFT_APB_WORD0);
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL25, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL25, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL26, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_7P5K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL28, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL28, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL29, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_17_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL31, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL31, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL32, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_7P5K_USEC);
-	} else {
-		preboot_puts("Invalid sram repair operation\n");
-		return -EINVAL;
-	}
-	regmap_writel(GAUDI2_RING_PATTERN_VAL33, mmPSOC_DFT_APB_WORD0);
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000010100000010100000010100;
-	expected_zeros = 0b00000000001011100001011100001011;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00001010000001010000001010000001;
-	expected_zeros = 0b10000101110000101110000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b01000000101000000101000000101000;
-	expected_zeros = 0b10111000010111000010111000010111;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST)
-		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
-	else
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00001010000001010000001010000000;
-	expected_zeros = 0b00000101110000101110000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL23, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL34, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000100100100100100100100100;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000000000010010010010010000;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-
-	}
-
-	return ret;
-}
-
-static int gaudi2_sob_rot_ring_pattern(enum gaudi2_sram_self_repair_op op)
-{
-	uint32_t expected_ones, expected_zeros;
-	int err, ret = 0;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL35, mmPSOC_DFT_APB_WORD0);
-
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL2, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_8K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL3, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_57_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL4, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_7P5K_USEC);
-	} else {
-		preboot_puts("Invalid sram repair operation\n");
-		return -EINVAL;
-	}
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL36, mmPSOC_DFT_APB_WORD0);
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000010100000010100000010100;
-	expected_zeros = 0b00000000001011100001011100001011;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST)
-		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
-	else
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000000000000000001010000000;
-	expected_zeros = 0b00000000000000000010000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL35, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL37, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000000000000010010010010000;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-	}
-
-	return ret;
-}
-
 static int gaudi2_sram_ring_pattern(enum gaudi2_sram_self_repair_op op)
 {
 	uint32_t expected_ones, expected_zeros;
@@ -412,608 +219,6 @@
 	return ret;
 }
 
-static int gaudi2_xbar_mid_ring_pattern(enum gaudi2_sram_self_repair_op op)
-{
-	uint32_t expected_ones, expected_zeros;
-	int err, ret = 0;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL1, mmPSOC_DFT_APB_WORD0);
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL2, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_4K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL3, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_20_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL4, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_4K_USEC);
-	} else {
-		preboot_puts("Invalid sram repair operation\n");
-		return -EINVAL;
-	}
-	regmap_writel(GAUDI2_RING_PATTERN_VAL5, mmPSOC_DFT_APB_WORD0);
-	if (op == SRAM_SELF_REPAIR_OP_BIST)
-		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
-	else
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000000000000000001010000000;
-	expected_zeros = 0b00000000000000000000000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL1, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL10, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000000000000000000000010000;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-	}
-
-	return ret;
-}
-
-static int gaudi2_vdec_ring_pattern(enum gaudi2_sram_self_repair_op op)
-{
-	uint32_t expected_ones, expected_zeros;
-	int err, ret = 0;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL8, mmPSOC_DFT_APB_WORD0);
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_4K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_104_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_4K_USEC);
-	} else {
-		preboot_puts("Invalid sram repair operation\n");
-		return -EINVAL;
-	}
-	regmap_writel(GAUDI2_RING_PATTERN_VAL11, mmPSOC_DFT_APB_WORD0);
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000010100000010100000010100;
-	expected_zeros = 0b00000000001011100001011100001011;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST)
-		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
-	else
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000000000000000000000000000;
-	expected_zeros = 0b00000000000000000000000000010000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL8, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL5, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 00000000000000000000000000000000;
-		expected_zeros = 00000000000000000000010010010000;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-	}
-
-	return ret;
-}
-
-static int gaudi2_sob_xdma_ring_pattern(enum gaudi2_sram_self_repair_op op)
-{
-	uint32_t expected_ones, expected_zeros;
-	int err, ret = 0;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL38, mmPSOC_DFT_APB_WORD0);
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL26, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_5K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL29, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_36_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL32, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_5K_USEC);
-	} else {
-		preboot_puts("Invalid sram repair operation\n");
-		return -EINVAL;
-	}
-	regmap_writel(GAUDI2_RING_PATTERN_VAL39, mmPSOC_DFT_APB_WORD0);
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000010100000010100000010100;
-	expected_zeros = 0b00000000001011100001011100001011;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST)
-		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
-	else
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000000001010000001010000000;
-	expected_zeros = 0b00000000010000101110000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL38, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL40, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000000000010010010010010000;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-	}
-
-	return ret;
-}
-
-static int gaudi2_nic_ring_pattern(enum gaudi2_sram_self_repair_op op)
-{
-	uint32_t expected_ones, expected_zeros;
-	int err, ret = 0;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL41, mmPSOC_DFT_APB_WORD0);
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL25, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL2, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_8K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL28, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL3, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_44_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL31, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL4, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_7P5K_USEC);
-	} else {
-		preboot_puts("Invalid sram repair operation\n");
-		return -EINVAL;
-	}
-	regmap_writel(GAUDI2_RING_PATTERN_VAL42, mmPSOC_DFT_APB_WORD0);
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000010100000010100000010100;
-	expected_zeros = 0b00000000001011100001011100001011;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00001010000001010000001010000001;
-	expected_zeros = 0b10000101110000101110000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST)
-		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
-	else
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000000000010000001010000000;
-	expected_zeros = 0b00000000000000101110000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL41, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL8, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000100100100100100100100100;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL21, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000000000000000000000000000;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-	}
-
-	return ret;
-}
-
-static int gaudi2_mme_ring_pattern(enum gaudi2_sram_self_repair_op op)
-{
-	uint32_t expected_ones, expected_zeros;
-	int err, ret = 0;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL42, mmPSOC_DFT_APB_WORD0);
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL25, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL26, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_8K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL28, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL29, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_55_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL31, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL32, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_7P5K_USEC);
-	} else {
-		preboot_puts("Invalid sram repair operation\n");
-		return -EINVAL;
-	}
-	regmap_writel(GAUDI2_RING_PATTERN_VAL43, mmPSOC_DFT_APB_WORD0);
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000010100000010100000010100;
-	expected_zeros = 0b00000000001011100001011100001011;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00001010000001010000001010000001;
-	expected_zeros = 0b10000101110000101110000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST)
-		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
-	else
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000010000001010000001010000000;
-	expected_zeros = 0b00000101110000101110000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL42, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL11, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000100100100100100100100100;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000000000000000000000010000;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-	}
-
-	return ret;
-}
-
-static int gaudi2_hmmu_xbar_edge_mc_ring_pattern(enum gaudi2_sram_self_repair_op op)
-{
-	uint32_t expected_ones, expected_zeros;
-	int err, ret = 0;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL44, mmPSOC_DFT_APB_WORD0);
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL25, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL45, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_8K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL28, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL46, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_104_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL31, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL47, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_7P5K_USEC);
-	} else {
-		preboot_puts("Invalid sram repair operation\n");
-		return -EINVAL;
-	}
-	regmap_writel(GAUDI2_RING_PATTERN_VAL48, mmPSOC_DFT_APB_WORD0);
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000010100000010100000010100;
-	expected_zeros = 0b00000000001011100001011100001011;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00001010000001010000001010000001;
-	expected_zeros = 0b10000101110000101110000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b01000000101000000101000000101000;
-	expected_zeros = 0b10111000010111000010111000010111;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST)
-		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
-	else
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000000000000000000000000000;
-	expected_zeros = 0b00000000000000000000000000000000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL44, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL49, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000100100100100100100100100;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000000000000000000010010000;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-	}
-
-	return ret;
-}
-
-static int gaudi2_hmmu_2_ring_pattern(enum gaudi2_sram_self_repair_op op)
-{
-	uint32_t expected_ones, expected_zeros;
-	int err, ret = 0;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL50, mmPSOC_DFT_APB_WORD0);
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL26, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_8K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL29, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_104_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL32, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_7P5K_USEC);
-	} else {
-		preboot_puts("Invalid sram repair operation\n");
-		return -EINVAL;
-	}
-	regmap_writel(GAUDI2_RING_PATTERN_VAL51, mmPSOC_DFT_APB_WORD0);
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST)
-		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
-	else
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000000001010000001010000000;
-	expected_zeros = 0b00000000000000101110000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL50, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL52, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000000000000000000010010000;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-	}
-
-	return ret;
-}
-
-static int gaudi2_edma1_hmmu3_ring_pattern(enum gaudi2_sram_self_repair_op op)
-{
-	uint32_t expected_ones, expected_zeros;
-	int err, ret = 0;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL38, mmPSOC_DFT_APB_WORD0);
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL26, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_8K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL29, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_100K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL32, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_7P5K_USEC);
-	} else {
-		preboot_puts("Invalid sram repair operation\n");
-		return -EINVAL;
-	}
-	regmap_writel(GAUDI2_RING_PATTERN_VAL39, mmPSOC_DFT_APB_WORD0);
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000010100000010100000010100;
-	expected_zeros = 0b00000000001011100001011100001011;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST)
-		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
-	else
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000000001010000001010000000;
-	expected_zeros = 0b00000000010000101110000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL38, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL40, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000000000010010010010010000;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-	}
-
-	return ret;
-}
-
-static int gaudi2_hmmu_mc_ring_pattern(enum gaudi2_sram_self_repair_op op)
-{
-	uint32_t expected_ones, expected_zeros;
-	int err, ret = 0;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL53, mmPSOC_DFT_APB_WORD0);
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL24, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL25, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_8K_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BISR) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL27, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL28, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_104_USEC);
-	} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL30, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL31, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-		udelay(DELAY_7P5K_USEC);
-	} else {
-		preboot_puts("Invalid sram repair operation\n");
-		return -EINVAL;
-	}
-	regmap_writel(GAUDI2_RING_PATTERN_VAL55, mmPSOC_DFT_APB_WORD0);
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00000000010100000010100000010100;
-	expected_zeros = 0b00000000001011100001011100001011;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-
-	expected_ones = 0b00001010000001010000001010000001;
-	expected_zeros = 0b10000101110000101110000101110000;
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST)
-		regmap_writel(GAUDI2_RING_PATTERN_VAL6, mmPSOC_DFT_APB_WORD1);
-	else
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-	if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-		expected_ones = 0b00000000000001010000001010000000;
-		expected_zeros = 0b00000000010000101110000101110000;
-	} else {
-		expected_ones = 0b00000000000000000000000010000000;
-		expected_zeros = 0b00000000000000000000000101110000;
-	}
-	err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-	if (err)
-		ret = 1;
-
-	if (op == SRAM_SELF_REPAIR_OP_BIST) {
-		regmap_writel(GAUDI2_RING_PATTERN_VAL53, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL0, mmPSOC_DFT_APB_WORD1);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
-
-		regmap_writel(GAUDI2_RING_PATTERN_VAL54, mmPSOC_DFT_APB_WORD0);
-		regmap_writel(GAUDI2_RING_PATTERN_VAL7, mmPSOC_DFT_APB_WORD1);
-
-		expected_ones = 0b00000000000000000000000000000000;
-		expected_zeros = 0b00000000010010010010010010010000;
-		err = gaudi2_sms_read_and_check_pattern(expected_ones, expected_zeros);
-		if (err)
-			ret = 1;
-	}
-
-	return ret;
-}
-
 static void gaudi2_sram_ring_sel(uint32_t dfx_val, uint32_t ring)
 {
 	uint32_t ring_value = BIT(8) | (ring << 3) | 0x5;
@@ -1051,100 +256,49 @@
 	regmap_writel(GAUDI2_RING_PATTERN_VAL9, mmPSOC_DFT_APB_WORD1);
 }
 
-static int (*gaudi2_sram_region_ring_pattern[])(enum gaudi2_sram_self_repair_op op)  = {
-				[SRAM_RING_TYPE_TPC] = gaudi2_tpc_ring_pattern,
-				[SRAM_RING_TYPE_SOB_ROT] = gaudi2_sob_rot_ring_pattern,
-				[SRAM_RING_TYPE_SRAM] = gaudi2_sram_ring_pattern,
-				[SRAM_RING_TYPE_XBAR_MID] = gaudi2_xbar_mid_ring_pattern,
-				[SRAM_RING_TYPE_VDEC] = gaudi2_vdec_ring_pattern,
-				[SRAM_RING_TYPE_SOB_XDMA] = gaudi2_sob_xdma_ring_pattern,
-				[SRAM_RING_TYPE_NIC] = gaudi2_nic_ring_pattern,
-				[SRAM_RING_TYPE_MME] = gaudi2_mme_ring_pattern,
-				[SRAM_RING_TYPE_HMMU_XBAR] = gaudi2_hmmu_xbar_edge_mc_ring_pattern,
-				[SRAM_RING_TYPE_HMMU_2] = gaudi2_hmmu_2_ring_pattern,
-				[SRAM_RING_TYPE_EDMA1_HMMU3] = gaudi2_edma1_hmmu3_ring_pattern,
-				[SRAM_RING_TYPE_HMMU_MC] = gaudi2_hmmu_mc_ring_pattern
-};
-
 #define GAUDI2_DFX_SHIFT_VAL		5
 static void gaudi2_sram_self_repair(enum gaudi2_sram_self_repair_op op)
 {
-	unsigned int dfx, ring;
-	enum gaudi2_sram_ring_type ring_type;
+	unsigned int ring;
 	int ret;
 	u8 val, num;
 
 	gaudi2_sram_beginning_pattern();
 
-	for (dfx = 0; dfx < ARRAY_SIZE(dfx_value_table); dfx++) {
-		for (ring = 0; ring < ring_count_table[dfx]; ring++) {
-			ring_type = gauid2_ring_type_table[dfx][ring];
-			if (ring_type != SRAM_RING_TYPE_MAX) {
-				gaudi2_sram_id_sel();
-				gaudi2_sram_ring_sel(dfx_value_table[dfx], ring + 1);
-				ret = gaudi2_sram_region_ring_pattern[ring_type](op);
-				if (ret > 0) {
-					val = (dfx << GAUDI2_DFX_SHIFT_VAL) | (ring + 1);
-					if (op == SRAM_SELF_REPAIR_OP_BIRA) {
-						num = bist_data->num_prebist_failed_rings;
-						bist_data->pre_bist_failed_rings[num] = val;
-						bist_data->num_prebist_failed_rings++;
-						preboot_puts_hex("bira failed val: ", val, 0);
-					} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
-						num = bist_data->num_postbist_failed_rings;
-						bist_data->post_bist_failed_rings[num] = val;
-						bist_data->num_postbist_failed_rings++;
-						preboot_puts_hex("bist failed val: ", val, 0);
-					} else {
-						preboot_puts_hex("bisr failed val: ", val, 0);
-					}
-				}
-				gaudi2_sram_id_sel();
+	for (ring = 0; ring < GAUDI2_SRAM_RINGS_NUM; ring++) {
+		gaudi2_sram_id_sel();
+		gaudi2_sram_ring_sel(dfx_value_table[gaudi2_sram_rings[ring].dfx],
+								gaudi2_sram_rings[ring].ring);
+		ret = gaudi2_sram_ring_pattern(op);
+		if (ret > 0) {
+			val = (gaudi2_sram_rings[ring].dfx << GAUDI2_DFX_SHIFT_VAL) |
+								gaudi2_sram_rings[ring].ring;
+			if (op == SRAM_SELF_REPAIR_OP_BIRA) {
+				num = bist_data->num_prebist_failed_rings;
+				bist_data->pre_bist_failed_rings[num] = val;
+				bist_data->num_prebist_failed_rings++;
+				preboot_puts_hex("bira failed val: ", val, 0);
+			} else if (op == SRAM_SELF_REPAIR_OP_BIST) {
+				num = bist_data->num_postbist_failed_rings;
+				bist_data->post_bist_failed_rings[num] = val;
+				bist_data->num_postbist_failed_rings++;
+				preboot_puts_hex("bist failed val: ", val, 0);
+			} else {
+				preboot_puts_hex("bisr failed val: ", val, 0);
 			}
 		}
+		gaudi2_sram_id_sel();
 	}
 }
 
-static void gaudi2_reset_engines(void)
+static void gaudi2_reset_srams(void)
 {
-	u32 tpc_iso_val, hbm_iso_val, nic_iso_val;
-
-	tpc_iso_val = regmap_readl(mmPSOC_GLOBAL_CONF_TPC_ISO);
-	regmap_writel(0, mmPSOC_GLOBAL_CONF_TPC_ISO);
-	hbm_iso_val = regmap_readl(mmPSOC_GLOBAL_CONF_HBM_ISO);
-	regmap_writel(0, mmPSOC_GLOBAL_CONF_HBM_ISO);
-	nic_iso_val = regmap_readl(mmPSOC_GLOBAL_CONF_NIC_ISO);
-	regmap_writel(0, mmPSOC_GLOBAL_CONF_NIC_ISO);
-
-	regmap_writel(PSOC_RESET_CONF_TPC_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_TPC_UNIT_RST);
-	regmap_writel(PSOC_RESET_CONF_TPC_DIV_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_TPC_DIV_UNIT_RST);
-	regmap_writel(PSOC_RESET_CONF_ROTATOR_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_ROTATOR_UNIT_RST);
-	regmap_writel(PSOC_RESET_CONF_SM_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_SM_UNIT_RST);
-	regmap_writel(PSOC_RESET_CONF_HBM_MC_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_HBM_MC_UNIT_RST);
-	regmap_writel(PSOC_RESET_CONF_MME_EUS_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_MME_EUS_UNIT_RST);
-	regmap_writel(PSOC_RESET_CONF_MSS_CLS_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_MSS_CLS_UNIT_RST);
-	regmap_writel(PSOC_RESET_CONF_NIC_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_NIC_UNIT_RST);
-	regmap_writel(PSOC_RESET_CONF_NIC_PRT_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_NIC_PRT_UNIT_RST);
 	regmap_writel(PSOC_RESET_CONF_SRAM_UNIT_RST_EN_MASK, mmPSOC_RESET_CONF_SRAM_UNIT_RST);
 	regmap_writel(PSOC_RESET_CONF_UNIT_RST_N_IND_MASK, mmPSOC_RESET_CONF_UNIT_RST_N);
 	udelay(DELAY_1K_USEC); /* TODO: confirm the delay */
 	regmap_writel(0, mmPSOC_RESET_CONF_UNIT_RST_N);
-	regmap_writel(0, mmPSOC_RESET_CONF_TPC_UNIT_RST);
-	regmap_writel(0, mmPSOC_RESET_CONF_TPC_DIV_UNIT_RST);
-	regmap_writel(0, mmPSOC_RESET_CONF_SM_UNIT_RST);
-	regmap_writel(0, mmPSOC_RESET_CONF_ROTATOR_UNIT_RST);
-	regmap_writel(0, mmPSOC_RESET_CONF_HBM_MC_UNIT_RST);
-	regmap_writel(0, mmPSOC_RESET_CONF_MME_EUS_UNIT_RST);
-	regmap_writel(0, mmPSOC_RESET_CONF_MSS_CLS_UNIT_RST);
-	regmap_writel(0, mmPSOC_RESET_CONF_NIC_UNIT_RST);
-	regmap_writel(0, mmPSOC_RESET_CONF_NIC_PRT_UNIT_RST);
 	regmap_writel(0, mmPSOC_RESET_CONF_SRAM_UNIT_RST);
 	udelay(DELAY_1K_USEC); /* TODO: confirm the delay */
-
-	/* Restore original iso value */
-	regmap_writel(tpc_iso_val, mmPSOC_GLOBAL_CONF_TPC_ISO);
-	regmap_writel(hbm_iso_val, mmPSOC_GLOBAL_CONF_HBM_ISO);
-	regmap_writel(nic_iso_val, mmPSOC_GLOBAL_CONF_NIC_ISO);
 }
 
 static void init_sram_bist_data(void)
@@ -1175,6 +329,6 @@
 	gaudi2_sram_self_repair(SRAM_SELF_REPAIR_OP_BIST);
 	/* Get back to normal from DFT mode */
 	regmap_writel(DFT_APB_CONTROL_DEFAULT_VAL, mmPSOC_GLOBAL_CONF_DFT_APB_CONTROL);
-	gaudi2_reset_engines();
-	preboot_puts("sram bist done\n");
+	gaudi2_reset_srams();
+	preboot_puts("bist done for 48MB SRAM\n");
 }
